tests/protobuf-net.Grpc.Reflection.Test/ReflectionServiceTests.cs

```diff
@@ -0,0 +1,145 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Google.Protobuf.Reflection;
+using grpc.reflection.v1alpha;
+using ProtoBuf;
+using ProtoBuf.Grpc.Configuration;
+using ProtoBuf.Grpc.Reflection;
+using Xunit;
+
+namespace protobuf_net.Grpc.Reflection.Test
+{
+    using ReflectionTest;
+
+    public class ReflectionServiceTests
+    {
+        [Theory]
+        [MemberData(nameof(Dependencies))]
+        public async Task ShouldIncludeDependenciesInCorrectOrder(Type service, string symbolName, string[] expectedDescriptors)
+        {
+            IServerReflection reflectionService = new ReflectionService(service);
+
+            await foreach (var response in reflectionService.ServerReflectionInfoAsync(GetRequest()))
+            {
+                var fileDescriptors = response
+                    .FileDescriptorResponse
+                    .FileDescriptorProtoes
+                    .Select(x => Serializer.Deserialize<FileDescriptorProto>(x.AsSpan()))
+                    .ToArray();
+
+                Assert.Equal(expectedDescriptors, fileDescriptors.Select(x => x.Name));
+            }
+
+            async IAsyncEnumerable<ServerReflectionRequest> GetRequest()
+            {
+                yield return new ServerReflectionRequest
+                {
+                    FileContainingSymbol = symbolName
+                };
+
+                await Task.CompletedTask;
+            }
+        }
+
+        public static IEnumerable<object[]> Dependencies => new[]
+        {
+            new object[]
+            {
+                typeof(BclService),
+                ".ReflectionTest.BclService",
+                new[]
+                {
+                    "ProtoBuf.Grpc.Internal.Empty.proto", // TODO: Maybe Google well-known should have correct name?
+                    "protobuf-net/bcl.proto",
+                    "ReflectionTest.BclMessage.proto",
+                    "ReflectionTest.BclService.proto"
+                }
+            },
+            new object[]
+            {
+                typeof(Nested),
+                ".ReflectionTest.Nested",
+                new[]
+                {
+                    "ReflectionTest.Three.proto", // This includes Two and One
+                    "ReflectionTest.Four.proto", // This includes Three, Two and One
```

Comment:
These can become messy when file descriptors contains multiple messages... Especially if one would like to use the reflection response for e.g. generating `.proto` files.

Reply:
Btw, @mgravell, any reason we shouldn't switch to use `SchemaGenerator` instead in the `FileDescriptorSetFactory`? This would include everything in one FileDescriptor per service..

Reply:
I've eyeballed `FileDescriptorSetFactory` and I'm not entirely sure what you mean here; happy to consider changes, but: it isn't leaping out at me. Possibly I just need more coffee.

Reply:
They both do similar things, although `SchemaGenerator` only generates the `.proto` schema, Reflection service should return `FileDescriptors`. But we could use `SchemaGenerator` as base, and then let `protobuf-net.Reflection` convert it to a `FileDescriptor`. This way we would get one "self-contained" `FileDescriptor` per service (not really, since well-known types will be available as separate descriptors).

Reply:
I went the SchemaGenerator route